test/mpi/coll/allred.c

```diff
@@ -267,26 +267,26 @@ static void set_index_power(MPI_Datatype mpi_type, int category, void *arr, int
     }
 }
 
-#define SET_PAIR_CASE(type, mpi_type) \
-    case mpi_type: \
+#define SET_PAIR_CASE(type, mpi_type_) \
+    if ( mpi_type == mpi_type_ ) \
         { \
             struct type ## _test *pp = p; \
             pp->a = val_a; \
             pp->b = val_b; \
-        }; \
-        break
+        } \
 
 static void set_pair_val(MPI_Datatype mpi_type, void *p, int val_a, int val_b)
 {
-    switch (mpi_type) {
-            SET_PAIR_CASE(int, MPI_2INT);
-            SET_PAIR_CASE(long, MPI_LONG_INT);
-            SET_PAIR_CASE(short, MPI_SHORT_INT);
-            SET_PAIR_CASE(float, MPI_FLOAT_INT);
-            SET_PAIR_CASE(double, MPI_DOUBLE_INT);
-        default:
-            assert(0);
-    }
+        SET_PAIR_CASE(int, MPI_2INT)
+        else
+        SET_PAIR_CASE(long, MPI_LONG_INT)
+        else
+        SET_PAIR_CASE(short, MPI_SHORT_INT)
+        else
+        SET_PAIR_CASE(float, MPI_FLOAT_INT)
+        else
+        SET_PAIR_CASE(double, MPI_DOUBLE_INT)
+        else { assert(0); }
```

Comment:
This is ugly :) I would rather spell the macros out. There are only 5 cases anyway. Curiously, is there a system missing one of these type? `int`, `long`, `short`, `float`, `double` are standard C types, right?

Reply:
Alternatively, define the macro for the case body, then
```
if (mpi_type == MPI_2INT) {
    SET_PAIR(int);
} else if (mpi_type == MPI_LONG_INT) {
    SET_PAIR(long);
} ...
```
Rationale: the `if-else if - else` is a semantic unit, it is bad to split in and out of macros.

Reply:
Also, the spell checker doesn't work with macros that can't follow `;`.

Reply:
Sure, I was merely trying to pattern match the previous version as much as possible.  I'll go with the `SET_PAIR` macro idea above.